server/src/main/java/org/elasticsearch/index/translog/TranslogHeader.java

```diff
@@ -0,0 +1,164 @@
+    }
+
+    private static int headerSize(int version, int uuidLength) {
+        int size = CodecUtil.headerLength(TRANSLOG_CODEC);
+        size += Integer.BYTES + uuidLength; // uuid
+        if (version >= VERSION_PRIMARY_TERM) {
+            size += Long.BYTES;    // primary term
+            size += Integer.BYTES; // checksum
+        }
+        return size;
+    }
+
+    /**
+     * Read a translog header from the given path and file channel
+     */
+    static TranslogHeader read(final String translogUUID, final Path path, final FileChannel channel) throws IOException {
+        // This input is intentionally not closed because closing it will close the FileChannel.
+        final BufferedChecksumStreamInput in =
+            new BufferedChecksumStreamInput(new InputStreamStreamInput(Channels.newInputStream(channel), channel.size()));
+        final int version;
+        try {
+            version = CodecUtil.checkHeader(new InputStreamDataInput(in), TRANSLOG_CODEC, VERSION_CHECKPOINTS, VERSION_PRIMARY_TERM);
+        } catch (CorruptIndexException | IndexFormatTooOldException | IndexFormatTooNewException e) {
+            throw new TranslogCorruptedException("Translog header corrupted. path:" + path, e);
+        }
+        // Read the translogUUID
+        final int uuidLen = in.readInt();
+        if (uuidLen > channel.size()) {
+            throw new TranslogCorruptedException("uuid length can't be larger than the translog");
+        }
+        final BytesRef uuid = new BytesRef(uuidLen);
+        uuid.length = uuidLen;
+        in.read(uuid.bytes, uuid.offset, uuid.length);
+        final BytesRef expectedUUID = new BytesRef(translogUUID);
+        if (uuid.bytesEquals(expectedUUID) == false) {
+            throw new TranslogCorruptedException("expected shard UUID " + expectedUUID + " but got: " + uuid +
+                " this translog file belongs to a different translog. path:" + path);
+        }
+        // Read the primary term
+        final long primaryTerm;
+        if (version == VERSION_PRIMARY_TERM) {
+            primaryTerm = in.readLong();
+            assert primaryTerm >= 0 : "Primary term must be non-negative [" + primaryTerm + "]; translog path [" + path + "]";
+        } else {
+            assert version == VERSION_CHECKPOINTS : "Unknown header version [" + version + "]";
+            primaryTerm = UNKNOWN_PRIMARY_TERM;
+        }
+        // Verify the checksum
+        if (version >= VERSION_PRIMARY_TERM) {
+            Translog.verifyChecksum(in);
+        }
+        final int headerSizeInBytes = headerSize(version, uuid.length);
+        assert channel.position() == headerSizeInBytes :
+            "Header is not fully read; header size [" + headerSizeInBytes + "], position [" + channel.position() + "]";
+        return new TranslogHeader(translogUUID, primaryTerm, headerSizeInBytes);
+    }
+
+    /**
+     * Writes this header with the latest format into the file channel
+     */
```

Comment:
check if you need to put a SuppressWarning here since we don't close the channel.

Reply:
Yes. I added a SuppressWarning for both Eclipse and Intellij.